src/v/resource_mgmt/rate.h

```diff
@@ -99,3 +101,85 @@ class rate_tracker final {
     clock::duration _window_size;
     std::size_t _current;
 };
+
+/// Token bucket algorithm, useful for rate-limiting solutions.
+///
+/// Keeps track of a limited number of resources known as 'tokens'.
+/// The current amount of tokens are kept in \ref bucket and when a user
+/// consumes resources they take a corresponding number of tokens from the
+/// bucket. Every \ref window duration the pre-configured amount of units are
+/// added to the bucket. If there are enough tokens in the bucket when queried
+/// the algorithm recommends rate-limiting, otherwise not.
+class token_bucket_rate_tracker {
+public:
+    using clock = ss::lowres_clock;
+    using units_t = int64_t;
+    using recommend_rate_limit
+      = ss::bool_class<struct recommend_rate_limit_tag>;
+
+    /// Class constructor
+    /// @param quota Expected rate to operate within
+    /// @param samples Number of windows to accumulate in memory for burst
+    /// @param window_sec width of window in which tokens are added - in seconds
+    token_bucket_rate_tracker(
+      uint32_t quota, uint32_t samples, clock::duration window_sec) noexcept
+      : _quota(quota)
+      , _samples(samples)
+      , _window(window_sec)
+      , _last_check() {}
+
+    /// Slight variation of token bucket algorithm to allow bursty traffic
+    /// \ref burst tokens are removed from bucket after returning throttle
+    /// behavior suggestion
+    recommend_rate_limit
+    record_and_measure(uint32_t n, const clock::time_point& now) noexcept {
+        refill(now);
+        if (_bucket < 0) {
+            return recommend_rate_limit::yes;
+        }
+        _bucket = std::min<units_t>(burst(), _bucket - n);
+        return recommend_rate_limit::no;
+    }
+
+    void record(uint32_t n, const clock::time_point& now) noexcept {
+        refill(now);
+        if (_bucket >= 0) {
+            /// Number of tokens in bucket is allowed to become negative to
+            /// allow burst of traffic to proceed
+            _bucket = std::min<units_t>(burst(), _bucket - n);
+        }
+    }
+
+    units_t measure(const clock::time_point& now) noexcept {
+        refill(now);
+        return _bucket;
+    }
+
+    /// Returns the number of tokens that current remain in the bucket
+    units_t units() const noexcept { return _bucket; }
+
+private:
+    void refill(const clock::time_point& now_ms) noexcept {
+        const auto delta = std::chrono::duration_cast<std::chrono::seconds>(
```

Comment:
I did it this way because I think the algorithm necessitates that seconds are used here when refilling the number of tokens in the bucket due to the fact that the rate is expressed in terms of the window width, which is expected to be a duration in seconds.

Reply:
I don't think it requires integer seconds for the calculation, though. What happens when the last check is always less than a second ago?

Reply:
So in this case, then the next window hasn't appeared so `refill()` will not add more tokens to the bucket. And if `record()` was called, the total number of tokens will then be reduced by `n`, so that checks out